zookeeper/source/zk_nn_client.cc

```diff
@@ -369,19 +370,45 @@ bool ZkNnClient::add_block(AddBlockRequestProto &req,
 
   std::uint64_t block_id;
   auto data_nodes = std::vector<std::string>();
+  uint64_t block_group_id;
+  std::vector<char> block_indices;
 
   if (!znode_data.isEC) {
       add_block(file_path, block_id, data_nodes, replication_factor);
   } else {  // case when some EC policy is used.
-      // TODO(Nate): generate a block group and each part of a block group.
-      // TODO(Nate): use the hierarchical naming scheme.
+      add_block_group(
+          file_path, block_group_id, data_nodes, block_indices,
+          DEFAULT_PARITY_UNITS + DEFAULT_DATA_UNITS);
   }
 
+  // TODO(nate): this might be seriously wrong.
   block->set_offset(0);  // TODO(2016): Set this
   block->set_corrupt(false);
 
   buildExtendedBlockProto(block->mutable_b(), block_id, block_size);
 
+  // Populate optional fields for an EC block.
+  // i.e. block indices and storage IDs.
+  if (znode_data.isEC) {
+    // Add storage IDs for an EC block.
+    for (int i = 0; i < DEFAULT_DATA_UNITS + DEFAULT_PARITY_UNITS; i++) {
+      block->set_storageids(i, DEFAULT_STORAGE_ID);
+    }
+
+    // Add block indices for an EC block.
+    // Each byte (i.e. char) represents an index into the group.
+    std::string block_index_string;
+    for (int i = 0; i < DEFAULT_DATA_UNITS + DEFAULT_PARITY_UNITS; i++) {
+      block_index_string.push_back(block_indices[i]);
```

Comment:
It is also not clear to me that this field is populated. I think we should save the block index in the block metadata in the block groups metadata zookeeper subtree.

Reply:
this field = ```block_indices```? The logic that populates ```block_indices``` was omitted in the last commit. I am making a new commit that contains this change.